Python/Module5_OddsAndEnds/Writing_Good_Code.ipynb

```diff
@@ -0,0 +1,1028 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### What is It Good For? (Absolutely Nothing)\n",
+    "\n",
+    "As far as the Python interpreter is concerned, **type-hints have no impact on your code other than adding documentation**. That is, Python **will not enforce any type-checking based on your type-hints**. This is in stark contrast to strongly-typed languages, like C++, where you are required specify input and output types for a function as constraints that are strictly enforced by the compiler. \n",
+    "\n",
+    "Indeed, in Python we are free to pass whatever we want to our function regardless of its annotated type-hints. Let's pass an empty list, instead of a string, to the  `count_vowels` function that we defined above.\n",
+    "\n",
+    "```python\n",
+    "# Type-hints do not actually enforce type-checking.\n",
+    "# Here we pass an empty list instead of a string without \n",
+    "# any warnings/errors being raised\n",
+    ">>> count_vowels([])\n",
+    "0\n",
+    "```\n",
+    "\n",
+    "See that our function pays no notice to the fact that we violated our type-hint, which suggests that we will be passing in a string. It happily loops over the empty list and sums over zeros entries, thus the function returns `0`. This is why these annotations are merely called type *hints*, and not type *requirements*.\n",
+    "\n",
+    "While the CPython interpreter will never require or enforce type-hints during the lifetime of Python 3, there are 3rd party libraries and tooling that makes keen use of these annotations. *It is these 3rd-party capabilities that make type-hinting worth using*. [IDEs like PyCharm and VSCode](https://www.pythonlikeyoumeanit.com/Module1_GettingStartedWithPython/Getting_Started_With_IDEs_and_Notebooks.html) can leverage type hints to great effect. They will warn you when you pass objects to your function that do not abide by the type hints, and they will also leverage these annotations to keep track of the objects that are returned by the function. \n",
+    "\n",
+    "For example, suppose that we are using the `count_vowels` function in a library of code that we are developing, and we mistakenly think that the output of this function is a string that we want to capitalize. Our IDE will see that we are trying to capitalize an integer and will thus warn us:\n",
+    "\n",
+    "![PyCharm uses type-hints to check our code for consistency](pics/pycharm1.png)\n",
+    "\n",
+    "This saves us the trouble of having to run our code, hit an error, read through the stack trace, and debug our mistake. Instead we immediately see our inconsistency get flagged. In fact, we probably never would have even made the mistake at all as our IDE would have shown a list of integer-methods for us to choose from, and not string-methods. This would have likely alerted us to our misapprehension. \n",
```

Comment:
Instead**,** we immediately see our inconsistency

...never would have made the mistake at all**,** as our IDE would have...